subgraphs/notional-finance/src/mappings/notional.ts

```diff
@@ -0,0 +1,333 @@
+} from "../common/constants";
+import { bigIntToBigDecimal } from "../common/numbers";
+import { getOrCreateAccount } from "../getters/account";
+import { getOrCreateMarket, getMarketsWithStatus } from "../getters/market";
+import {
+  createBorrow,
+  createDeposit,
+  createLiquidate,
+  createRepay,
+  createWithdraw,
+} from "../setters/transactions";
+import {
+  getOrCreateAsset,
+  updateAccountAssetOnEmptyPortfolio,
+  updateAccountAssets,
+} from "../getters/accountAssets";
+import { getTokenFromCurrency } from "../common/util";
+import { addToArrayAtIndex, removeFromArrayAtIndex } from "../common/arrays";
+import { getOrCreateInterestRate } from "../getters/InterestRate";
+import { getOrCreateERC1155Token } from "../getters/token";
+
+export function handleLendBorrowTrade(event: LendBorrowTrade): void {
+  // params
+  const currencyId = event.params.currencyId;
+  const maturity = event.params.maturity;
+  const marketId = currencyId.toString() + "-" + maturity.toString();
+
+  // entities
+  const market = getOrCreateMarket(event, marketId);
+  const account = getOrCreateAccount(event.params.account.toHexString());
+  const token = getTokenFromCurrency(event, currencyId.toString());
+
+  // protocol contract
+  const notional = Notional.bind(Address.fromString(PROTOCOL_ID));
+
+  // update input token price
+  market.inputTokenPriceUSD = token.lastPriceUSD!;
+
+  // update liquidation params
+  const rateStorageCall = notional.try_getRateStorage(currencyId);
+  if (rateStorageCall.reverted) {
+    log.error(
+      "[handleLendBorrowTrade] getRateStorage for currencyId {} reverted",
+      [currencyId.toString()]
+    );
+  } else {
+    market.maximumLTV = BigDecimal.fromString(
+      (rateStorageCall.value.getEthRate().haircut * 0.01).toString()
+    );
+    market.liquidationThreshold = BigDecimal.fromString(
+      (rateStorageCall.value.getEthRate().rateBuffer * 0.01).toString()
+    );
+    market.liquidationPenalty = BigDecimal.fromString(
+      (
+        rateStorageCall.value.getEthRate().liquidationDiscount - INT_HUNDRED
+      ).toString()
+    );
+  }
+
+  // update output token
+  const encodedIdCall = notional.try_encodeToId(currencyId, maturity, 1);
+  if (encodedIdCall.reverted) {
+    log.error(
+      "[handleLendBorrowTrade] encodeToId for currencyId {}, maturity {} reverted",
+      [currencyId.toString(), maturity.toString()]
+    );
+  } else {
+    market.outputToken = getOrCreateERC1155Token(
+      PROTOCOL_ID,
+      encodedIdCall.value
+    ).id;
+  }
+
+  market.save();
+
+  // track status of markets
+  const allMarkets = getMarketsWithStatus(event);
+  if (allMarkets.activeMarkets.indexOf(market.id) < 0) {
+    allMarkets.activeMarkets = addToArrayAtIndex(
+      allMarkets.activeMarkets,
+      market.id,
+      0
+    );
+    allMarkets.save();
+  }
+
+  // get active markets
+  const currencyIds = [1, 2, 3, 4];
+  let activeMarkets: string[] = [];
+  for (let i = 0; i < currencyIds.length; i++) {
```

Comment:
this does not need to be changed, but I would recommend something along the lines of a while loop that continues to call `getActiveMarkets(i)` until it fails in the case they add more markets

Reply:
The looping here is on the `currencyIds` (rather than the `markets`) to fetch active markets associated with that currency. Unfortunately, there is no contract call that would allow us to fetch the `currencyIds` and loop on it.

Reply:
But there could be a while loop that starts at 1 and continues to call `try_getActiveMarkets(number)` until the call reverts. That is how you would know you covered all of the markets.